### PR TITLE
Restructure optimisers to enable passing of kwargs

### LINKED ISSUE #236: Pass kwargs to SciPy optimisers
### Feature description

At the moment, the SciPy optimisers aren't passed kwargs. As such, hyper parameter tuning is not available to the end user. This issue is to update the SciPy implementation to enable optimiser method, options, etc. changes via kwargs.

### Motivation

End users should be able to modify the SciPy optimisers, links to #195.

### Possible implementation

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by NicolaCourtier**
# Description

Enable valid keyword arguments to be passed through the set-up of an `Optimisation` class and restructure the optimiser classes so that all PINTS and SciPy optimisers are subclasses of the same BaseOptimiser.

## Issue reference
Fixes #236.

## Review
Before you mark your PR as ready for review, please ensure that you've considered the following:
- Updated the [CHANGELOG.md](https://github.com/pybop-team/PyBOP/blob/develop/CHANGELOG.md) in reverse chronological order (newest at the top) with a concise description of the changes, including the PR number.
- Noted any breaking changes, including details on how it might impact existing functionality.

## Type of change
- [x] New Feature: A non-breaking change that adds new functionality.
- [ ] Optimization: A code change that improves performance.
- [x] Bug Fix: A non-breaking change that addresses an issue.
- [ ] Documentation: Updates to documentation or new documentation for new features.
- [x] Refactoring: Non-functional changes that improve the codebase.
- [x] Style: Non-functional changes related to code style (formatting, naming, etc).
- [x] Testing: Additional tests to improve coverage or confirm functionality.
- [ ] Other: (Insert description of change)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybop-team/PyBOP/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All unit tests pass: `$ nox -s tests`
- [x] The documentation builds: `$ nox -s docs`

You can run integration tests, unit tests, and doctests together at once, using `$ nox -s quick`.

## Further checks:
- [x] Code is well-commented, especially in complex or unclear areas.
- [x] Added tests that prove my fix is effective or that my feature works.
- [x] Checked that coverage remains or improves, and added tests if necessary to maintain or increase coverage.

Thank you for contributing to our project! Your efforts help us to deliver great software.

### CODE REVIEW

File: pybop/_optimisation.py [hunk 1]
```diff
@@ -180,10 +191,15 @@ def _run_pybop(self):
         final_cost : float
             The final cost associated with the best parameters.
         """
+        if "maxiter" in optimiser_kwargs.keys():
```
Comment (BradyPlanden) [inline pybop/_optimisation.py:194]: Isn't this already defined in lines 99-105 above?

File: pybop/optimisers/scipy_optimisers.py [hunk 2]
```diff
@@ -3,36 +3,63 @@
...
+    bounds=None,
+    method="Nelder-Mead",
+    tol=1e-5,
+    options=dict(),
+)
+DEFAULT_SCIPY_DIFFERENTIAL_EVOLUTION_OPTIONS = dict(
+    bounds=None,
+    strategy="best1bin",
+    popsize=15,
+    maxiter=1000,
```
Comment (BradyPlanden) [inline pybop/optimisers/scipy_optimisers.py:16]: Will this be used? If I understand the call structure correctly, this will always be overwritten from the `optimisation.set_max_iterations()` method.

**Review by BradyPlanden — COMMENTED** [review]
Well done @NicolaCourtier - thanks for the addition!

I've added a few comments for you to look at. One addition is to update `optimiser.needs_sensitivities` depending on `self.options["method"]`. If not added in this PR, we should open an issue to add gradient support for the scipy optimisers.

Also, we should probably align the name of the bounds variable between the pints and the scipy optimisers. This can also be done in a separate PR.

File: pybop/_optimisation.py [hunk 3]
```diff
@@ -95,54 +96,64 @@ def __init__(
...
+
+                self.optimiser = self.optimiser(bounds=self.bounds, **optimiser_kwargs)
 
             else:
                 raise ValueError("Unknown optimiser type")
 
         if self.pints:
+            if optimiser_kwargs:
+                raise ValueError(
+                    "Additional keyword arguments cannot currently be passed to PINTS optimisers."
```
[the remark this answers is not in the document]
Comment (NicolaCourtier) [reply pybop/_optimisation.py:113]: I had similar thoughts, let's implement this in a separate PR!

File: pybop/_optimisation.py (hunk 1 above)
Comment (NicolaCourtier) [reply pybop/_optimisation.py:194]: Lines 99-105 apply properties defined by kwargs passed on initialisation, e.g.
`optim = pybop.Optimisation(cost, optimiser, maxiter=10)`
Whereas lines 194-197 apply further properties passed on a particular call, e.g.
`optim.run(maxiter=10)`
Passing kwargs via the run function will overwrite existing properties. Does this make sense?

**Comment by NicolaCourtier** [comment]
This PR now includes changes to the PINTS interface as well as SciPy and introduces a true `BaseOptimiser`. Addressing @BradyPlanden's points above:

1. Added gradient support for the SciPy optimisers, via the `jac` option. Note that SciPy optimisers do not need the `needs_sensitivities` property.

2. The same name `bounds` is used for all optimisers, although more could be done to align the bounds type.

3.  Added `update_options` to the `BasePintsOptimiser` to unpack the `optimiser_kwargs` and try to apply them to the corresponding optimiser variables, with a final catch for unexpected keywords.

Note that this update nests the PINTS optimiser within the optimiser property of an `Optimisation` object... so I'm open to discussion around these changes.

File: examples/scripts/spm_NelderMead.py [hunk 4]
```diff
@@ -60,9 +60,9 @@ def noise(sigma):
     verbose=True,
     allow_infeasible_solutions=True,
     sigma0=0.05,
+    max_iterations=100,
+    max_unchanged_iterations=45,
```
Comment (BradyPlanden) [inline examples/scripts/spm_NelderMead.py:64]: Aligning with other examples.

```suggestion
    max_unchanged_iterations=20,
```

File: examples/scripts/spm_adam.py [hunk 5]
```diff
@@ -60,9 +60,9 @@ def noise(sigma):
     verbose=True,
     allow_infeasible_solutions=True,
     sigma0=0.05,
+    max_iterations=100,
+    max_unchanged_iterations=45,
```
Comment (BradyPlanden) [inline examples/scripts/spm_adam.py:64]: Same as above, 
```suggestion
    max_unchanged_iterations=20,
```

File: examples/scripts/spm_descent.py [hunk 6]
```diff
@@ -37,9 +37,12 @@
 problem = pybop.FittingProblem(model, parameters, dataset)
 cost = pybop.SumSquaredError(problem)
 optim = pybop.Optimisation(
-    cost, optimiser=pybop.GradientDescent, sigma0=0.022, verbose=True
+    cost,
+    optimiser=pybop.GradientDescent,
+    sigma0=0.022,
```
Comment (BradyPlanden) [inline examples/scripts/spm_descent.py:42]: This a good chance to align with results from: https://github.com/pybop-team/PyBOP/blob/a6461da16b985cb8bf69b62219e2d6835e881afd/examples/notebooks/optimiser_calibration.ipynb
```suggestion
    sigma0=0.011,
```

File: pybop/_optimisation.py [hunk 7]
```diff
@@ -73,441 +71,60 @@ def __init__(
...
+            The best parameter set found by the optimisation.
         final_cost : float
             The final cost associated with the best parameters.
         """
 
-        if self.pints:
-            x, final_cost = self._run_pints()
-        elif not self.pints:
-            x, final_cost = self._run_pybop()
+        x, final_cost = self.optimiser.run(**optimiser_kwargs)
```
Comment (BradyPlanden) [inline pybop/_optimisation.py:103]: I think it would be helpful to keep the API for run/optimise the same if possible. I.e. passing a cost function and initial conditions then kwargs for all other options. This really helps understand what the function requirements and interactions are.

File: pybop/optimisers/base_optimiser.py [hunk 8]
```diff
@@ -3,78 +3,109 @@ class BaseOptimiser:
     A base class for defining optimisation methods.
 
     This class serves as a template for creating optimisers. It provides a basic structure for
```
Comment (BradyPlanden) [inline pybop/optimisers/base_optimiser.py:5]:
```suggestion
    This class serves as a base class for creating optimisers. It provides a basic structure for
```

File: pybop/optimisers/base_optimiser.py [hunk 9]
```diff
@@ -3,78 +3,109 @@ class BaseOptimiser:
...
+        self._cost_function = None
+        self.x0 = x0
+        self.sigma0 = sigma0
         self.bounds = bounds
+
+        self.log = []
+        self._max_iterations = None
+        self.set_max_iterations()
+
+    def update_options(self, **optimiser_kwargs):
```
Comment (BradyPlanden) [inline pybop/optimisers/base_optimiser.py:33]: To align with standard set/get interfacing. Possibly worth adding a get_options as well.
```suggestion
    def set_options(self, **optimiser_kwargs):
```

File: pybop/optimisers/pints_optimisers.py [hunk 10]
```diff
@@ -1,8 +1,428 @@
...
+    def update_options(self, **optimiser_kwargs):
+        """
+        Update the optimiser options.
+
+        Parameters
+        ----------
+        **optimiser_kwargs : optional
+            Valid PINTS option keys and their values.
+        """
+        for key, value in optimiser_kwargs.items():
```
Comment (BradyPlanden) [inline pybop/optimisers/pints_optimisers.py:93]: I wonder if there is a better way to map the kwarg to the individual option. This is feels like it could get overwhelming if we start adding additional options.

File: pybop/optimisers/pints_optimisers.py [hunk 11]
```diff
@@ -1,8 +1,428 @@
...
+            (default: None).
+        """
+        if evaluations is not None:
+            evaluations = int(evaluations)
+            if evaluations < 0:
+                raise ValueError("Maximum number of evaluations cannot be negative.")
+        self._max_evaluations = evaluations
+
+
+class GradientDescent(BasePintsOptimiser):
```
Comment (BradyPlanden) [inline pybop/optimisers/pints_optimisers.py:425]: I think it would be good split the pints' optimisers into a separate file to reduce the length of this file.

File: pybop/optimisers/pints_optimisers.py [hunk 12]
```diff
@@ -1,8 +1,428 @@
...
+        self._boundaries = boundaries
+
+        # Convert x0 to PINTS vector
+        self._x0 = pints.vector(self.x0)
+
+        # PyBOP doesn't currently support the PINTS transformation class
+        self._transformation = None
+
+        # Create an instance of the PINTS optimiser class
+        self.pints_optimiser = pints_class(x0, sigma0, self._boundaries)
```
Comment (BradyPlanden) [inline pybop/optimisers/pints_optimisers.py:43]:
```suggestion
        self.pints_optimiser = pints_class(self._x0, sigma0, self._boundaries)
```

File: pybop/optimisers/pints_optimisers.py [hunk 13]
```diff
@@ -18,21 +438,34 @@ class GradientDescent(pints.GradientDescent):
...
+    **optimiser_kwargs : optional
+        Valid PINTS option keys and their values.
 
     See Also
     --------
     pints.GradientDescent : The PINTS implementation this class is based on.
     """
 
-    def __init__(self, x0, sigma0=0.1, bounds=None):
+    def __init__(self, x0=None, sigma0=0.1, bounds=None, **optimiser_kwargs):
```
Comment (BradyPlanden) [inline pybop/optimisers/pints_optimisers.py:449]: Should x0 be optional in these classes? The pints' class has it as a required argument. 

Also a good time to update the default sigma0 value for the GradientDescent class:


```suggestion
    def __init__(self, x0=None, sigma0=0.02, bounds=None, **optimiser_kwargs):
```

**Review by BradyPlanden — COMMENTED** [review]
I thought I would add a review on this even though it's not marked ready. It's a great step for the optimisers. There are a few areas I think we can improve, but thanks for the addition @NicolaCourtier!

File: pybop/_optimisation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (NicolaCourtier) [reply pybop/_optimisation.py]: This is now part of this PR

File: pybop/_optimisation.py (hunk 7 above)
Comment (NicolaCourtier) [reply pybop/_optimisation.py]: Outdated

File: pybop/optimisers/base_optimiser.py (hunk 8 above)
Comment (NicolaCourtier) [reply pybop/optimisers/base_optimiser.py]: Resolved in _optimisation.py

File: pybop/optimisers/pints_optimisers.py (hunk 10 above)
Comment (NicolaCourtier) [reply pybop/optimisers/pints_optimisers.py]: Outdated

File: pybop/optimisers/pints_optimisers.py (hunk 12 above)
Comment (NicolaCourtier) [reply pybop/optimisers/pints_optimisers.py]: Outdated

**Comment by NicolaCourtier** [comment]
In the latest changes, `Optimisation` becomes the base optimiser class with `BaseSciPyOptimiser` and `BasePintsOptimiser` subclasses, on which on the optimisers themselves are based. This simplifies the structure and avoids unnecessary nesting of the optimiser within another object.

I have included a `pybop.DefaultOptimiser` to replace the creation of an Optimisation object without a specific optimiser. I have avoided renaming files, but in future we should move `._optimisation.py` into the `optimisers` folder and rename `base_optimiser.py` as `base_pints_optimiser.py` to reflect the new structure.

The aim of this restructure is to allow keyword arguments to be passed in the same way to any kind of optimiser, e.g.
`optim = pybop.DefaultOptimiser(cost, max_iterations=100)`
An error is returned if one or more keywords are not recognised. Future developments can extend the handling of additional keywords via the `_set_options` functionality.

File: pybop/optimisers/pints_optimisers.py [hunk 14]
```diff
@@ -241,30 +210,33 @@ class CMAES(pints.CMAES):
...
-            self.boundaries = pints.RectangularBoundaries(
-                bounds["lower"], bounds["upper"]
-            )
-        else:
-            self.boundaries = None
+        super().__init__(cost, pints.CMAES, **optimiser_kwargs)
 
-        super().__init__(x0, sigma0, self.boundaries)
+
+class DefaultOptimiser(XNES):
```
Comment (BradyPlanden) [inline pybop/optimisers/pints_optimisers.py:238]: Instead of providing a `pybop.DefaultOptimiser` class, we can point users to the `pybop.Optimisation` API. We can discuss a bit more when I'm back, but I think that's a cleaner way to interact with PyBOP if you don't know what optimiser you want.

File: pybop/optimisers/scipy_optimisers.py [hunk 15]
```diff
@@ -1,85 +1,193 @@
...
+            Valid SciPy option keys and their values.
+
+        Returns
+        -------
+        optimiser_kwargs : dict
+            Remaining option keys and their values.
+        """
+        # Unpack nested values from SciPy options dictionary
+        if "options" in optimiser_kwargs.keys():
+            options_list = list(optimiser_kwargs["options"].keys())
```
Comment (BradyPlanden) [inline pybop/optimisers/scipy_optimisers.py:58]: Same suggestion as above (pybop/_optimisation.py: Line 87)

File: pybop/optimisers/scipy_optimisers.py [hunk 16]
```diff
@@ -1,85 +1,193 @@
...
+            for key in options_list:
+                if key not in optimiser_kwargs.keys():
+                    optimiser_kwargs[key] = optimiser_kwargs["options"].pop(key)
+                else:
+                    optimiser_kwargs["options"].pop(key)  # remove entry
+
+        # Keep max_iterations in preference to maxiter
```
Comment (BradyPlanden) [inline pybop/optimisers/scipy_optimisers.py:65]: To confirm, we are taking preference of kwarg "max_iterations" over "maxiter"? If so this might be clearer:
```suggestion
        # Overwrite maxiter variable with kwarg max_iterations value
```

File: pybop/optimisers/scipy_optimisers.py [hunk 17]
```diff
@@ -1,85 +1,193 @@
...
+
+    Parameters
+    ----------
+    **optimiser_kwargs : optional
+        Valid SciPy Minimize option keys and their values, For example:
         x0 : array_like
-            Initial guess for the parameters.
+            Initial position from which optimisation will start.
+        sigma0 : float
+            Initial step size or standard deviation depending on the optimiser.
```
Comment (BradyPlanden) [inline pybop/optimisers/scipy_optimisers.py:118]:
```suggestion
```

File: pybop/optimisers/scipy_optimisers.py [hunk 18]
```diff
@@ -111,79 +208,80 @@ def name(self):
...
         """
-
-        self.log = []
-
-        if x0 is not None:
+        if self.x0 is not None:
             print(
                 "Ignoring x0. Initial conditions are not used for differential_evolution."
             )
+        self.x0 = None
```
Comment (BradyPlanden) [inline pybop/optimisers/scipy_optimisers.py:254]:
```suggestion
            self.x0 = None
```

File: pybop/optimisers/scipy_optimisers.py [hunk 19]
```diff
@@ -111,79 +208,80 @@ def name(self):
...
+            ):
+                raise ValueError("Bounds must be specified for differential_evolution.")
+            bounds = [
+                (lower, upper)
+                for lower, upper in zip(self.bounds["lower"], self.bounds["upper"])
+            ]
+        else:
+            if not all(np.isfinite(value) for pair in self.bounds for value in pair):
+                raise ValueError("Bounds must be specified for differential_evolution.")
+            bounds = self.bounds
```
Comment (BradyPlanden) [inline pybop/optimisers/scipy_optimisers.py:277]: I think this can be completed so that bounds is retained as a class attribute. I would improve readability if we retain `self.bounds` by either modifying inplace or creating a copy. We can do this in a separate PR if however.

File: pybop/optimisers/scipy_optimisers.py [hunk 20]
```diff
@@ -201,18 +299,7 @@ def set_population_size(self, population_size=None):
...
-        """
-        Determines if the optimization algorithm requires gradient information.
-
-        Returns
-        -------
-        bool
-            False, indicating that gradient information is not required for differential evolution.
-        """
-        return False
+            self.popsize = population_size
```
Comment (BradyPlanden) [inline pybop/optimisers/scipy_optimisers.py:302]: I think this class definition can now move into the `Optimisation` class.

File: pybop/optimisers/scipy_optimisers.py [hunk 21]
```diff
@@ -1,85 +1,193 @@
...
 import numpy as np
 from scipy.optimize import differential_evolution, minimize
 
-from .base_optimiser import BaseOptimiser
+from pybop import Optimisation
 
+DEFAULT_SCIPY_MINIMIZE_OPTIONS = dict(
+    method="Nelder-Mead",
+    jac=False,
+    tol=1e-5,
```
Comment (BradyPlanden) [inline pybop/optimisers/scipy_optimisers.py:11]: If we rename this variable to `threshold` the SciPy and Pints' classes will be able to use the same kwarg.

**Review by BradyPlanden — COMMENTED** [review]
Thank you for these additions @NicolaCourtier! It looks great, I've left comments for you to take a look at. This with the addition of #309 is a great step towards a flexible optimisation API.

File: examples/notebooks/LG_M50_ECM/1-single-pulse-circuit-model.ipynb [hunk 22]
```diff
@@ -1680,7 +1679,7 @@
     }
    ],
    "source": [
-    "pybop.quick_plot(problem, parameter_values=x, title=\"Optimised Comparison\");"
```
[the remark this answers is not in the document]
Comment (NicolaCourtier) [reply examples/notebooks/LG_M50_ECM/1-single-pulse-circuit-model.ipynb:1683]: Ah! Thank you for explaining, I did not notice this!

File: pybop/_optimisation.py [hunk 23]
```diff
@@ -28,417 +38,174 @@ class Optimisation:
...
+        key_list = list(optimiser_kwargs.keys())
+        for key in key_list:
+            if key in ["x0", "bounds", "sigma0", "verbose"]:
+                self.__dict__.update({key: optimiser_kwargs.pop(key)})
+            elif key == "allow_infeasible_solutions":
+                self.allow_infeasible_solutions = self.set_allow_infeasible_solutions(
+                    optimiser_kwargs.pop(key)
+                )
+            elif key == "parallel":
+                self.set_parallel(optimiser_kwargs.pop(key))
```
[the remark this answers is not in the document]
Comment (NicolaCourtier) [reply pybop/_optimisation.py:96]: The `optimiser_kwargs.pop(key)` also removes the key and its value from the `optimiser_kwargs` dictionary. This inline modification to the dictionary was the motivation for using a list. The aim is to ensure that any unused arguments remain and therefore trigger an error, so this behaviour would also need to be replicated in order to implement this suggestion.

File: pybop/_optimisation.py [hunk 24]
```diff
@@ -1,25 +1,35 @@
...
 import pybop
 
+DEFAULT_OPTIMISER_OPTIONS = dict(
+    x0=None,
+    bounds=None,
+    sigma0=0.1,
+    verbose=False,
+    physical_viability=True,
+    allow_infeasible_solutions=True,
+    _max_iterations=None,
```
[the remark this answers is not in the document]
Comment (NicolaCourtier) [reply pybop/_optimisation.py:12]: Such a change should be consistent with similar modifications to `_unchanged_max_iterations`, `_max_evaluations`, and `_threshold`. I would like to understand better the reason for this choice in PINTS before changing.

File: pybop/_optimisation.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (NicolaCourtier) [reply pybop/_optimisation.py:96]: Not so relevant here, but necessary in the optimiser-specific `_set_options`, so I kept the same process here for consistency.

File: pybop/optimisers/scipy_optimisers.py (hunk 21 above)
Comment (NicolaCourtier) [reply pybop/optimisers/scipy_optimisers.py:11]: True, but keeping this keyword will help users who refer to the SciPy documentation. We could instead make `tol` and `threshold` synonyms (like `maxiter` and `max_iterations`) in a future PR.

File: pybop/_optimisation.py [hunk 25]
```diff
@@ -28,417 +38,174 @@ class Optimisation:
...
 
-        # Post-run statistics
-        self._evaluations = None
-        self._iterations = None
+        Returns
+        -------
+        optimiser_kwargs : dict
+            Remaining option keys and their values.
+        """
+        return optimiser_kwargs
```
[the remark this answers is not in the document]
Comment (NicolaCourtier) [reply pybop/_optimisation.py:118]: Perhaps? If an optimiser does not require additional options then this function provides the default behaviour. Any options that are not set during the initialisation of an optimiser will raise an error.

**Comment by NicolaCourtier** [comment]
Change to the scale in line 15 of tests/integration/test_parameterisations.py may be undone after #321 is merged.

File: pybop/optimisers/pints_optimisers.py (hunk 14 above)
Comment (BradyPlanden) [reply pybop/optimisers/pints_optimisers.py:238]: Updated in #309

File: pybop/optimisers/scipy_optimisers.py [hunk 26]
```diff
@@ -1,104 +1,193 @@
...
+                if np.isinf(cost):
+                    self.inf_count += 1
+                    cost = 1 + 0.9**self.inf_count  # for fake finite gradient
+                return cost
+        elif self._options["jac"] is True:
+
+            def cost_wrapper(x):
+                return self.cost.evaluateS1(x, minimising=self._minimising)
+        else:
+            raise ValueError(
```
Comment (BradyPlanden) [inline pybop/optimisers/scipy_optimisers.py:177]: This check should be completed within the `_set_up_optimiser` method.

File: tests/unit/test_optimisation.py [hunk 27]
```diff
@@ -84,33 +84,141 @@ def two_param_cost(self, model, two_parameters, dataset):
...
-        # Test setting population size
-        if optimiser_class in [pybop.SciPyDifferentialEvolution]:
-            with pytest.raises(ValueError):
-                opt.optimiser.set_population_size(-5)
+        else:
+            # Check and update initial values
+            assert optim.x0 == cost.x0
+            x0_new = np.array([0.6])
+            optim = optimiser(cost=cost, x0=x0_new)
+            assert optim.x0 == x0_new
```
Comment (BradyPlanden) [inline tests/unit/test_optimisation.py:210]: It's probably a good idea to `assert optim.x0 != cost.x0` here.

File: tests/unit/test_optimisation.py [hunk 28]
```diff
@@ -84,33 +84,141 @@ def two_param_cost(self, model, two_parameters, dataset):
...
 
-            # Correct value
-            opt.optimiser.set_population_size(5)
+        if optimiser in [pybop.SciPyMinimize]:
+            # Check a method that uses gradient information
+            optimiser(cost=cost, method="L-BFGS-B", jac=True)
```
Comment (BradyPlanden) [inline tests/unit/test_optimisation.py:214]: We should limit the max_iter condition for these tests. No need to do a full run if we are just checking unit values.

File: tests/unit/test_optimisation.py [hunk 29]
```diff
@@ -84,33 +84,141 @@ def two_param_cost(self, model, two_parameters, dataset):
...
+            # Check and update initial values
+            assert optim.x0 == cost.x0
+            x0_new = np.array([0.6])
+            optim = optimiser(cost=cost, x0=x0_new)
+            assert optim.x0 == x0_new
 
-            # Correct value
-            opt.optimiser.set_population_size(5)
+        if optimiser in [pybop.SciPyMinimize]:
+            # Check a method that uses gradient information
```
Comment (BradyPlanden) [inline tests/unit/test_optimisation.py:213]: Do we have an assertion on this? It seems like we are missing it here.

File: tests/unit/test_optimisation.py [hunk 30]
```diff
@@ -171,19 +304,49 @@ def test_halting(self, cost):
...
+        with pytest.raises(Exception, match="Callback error message"):
+            optim.run()
+        optim._callback = None
+
+        # Trigger optimiser error
+        def optimiser_error():
+            return "Optimiser error message"
+
+        optim.method.stop = optimiser_error
+        optim.run()
```
Comment (BradyPlanden) [inline tests/unit/test_optimisation.py:328]: Please add an assertion here.

File: examples/standalone/optimiser.py [hunk 31]
```diff
@@ -0,0 +1,90 @@
...
+class StandaloneOptimiser(Optimisation):
+    """
+    Defines an example standalone optimiser without a Cost.
+    """
+
+    def __init__(self, cost=None, **optimiser_kwargs):
+        # Define cost function
+        def cost(x):
+            x1, x2 = x
+            return (x1 - 2) ** 2 + (x2 - 4) ** 4
```
[the remark this answers is not in the document]
Comment (NicolaCourtier) [reply examples/standalone/optimiser.py:16]: I'm not sure what you mean? This StandaloneOptimiser can be run without an instance of the BaseCost. Without these lines defining the cost function, it cannot be run independently.

File: pybop/plotting/plot_convergence.py [hunk 32]
```diff
@@ -30,23 +30,29 @@ def plot_convergence(optim, show=True, **layout_kwargs):
...
+    if optim._minimising:
+        best_cost_per_iteration = [
+            min((cost(solution) for solution in log_entry), default=np.inf)
+            for log_entry in log
+        ]
+    else:
+        best_cost_per_iteration = [
+            max((cost(solution) for solution in log_entry), default=-np.inf)
+            for log_entry in log
+        ]
```
[the remark this answers is not in the document]
Comment (NicolaCourtier) [reply pybop/plotting/plot_convergence.py:43]: I'm not sure that these changes will work. The if statement if used to employ min in the first instance and max in the second instance. The replacement lines are always minimising?

File: tests/integration/test_optimisation_options.py [hunk 33]
```diff
@@ -81,11 +81,8 @@ def spm_costs(self, model, parameters, cost_class):
     @pytest.mark.integration
     def test_optimisation_f_guessed(self, f_guessed, spm_costs):
         # Test each optimiser
-        parameterisation = pybop.Optimisation(
-            cost=spm_costs, optimiser=pybop.XNES, sigma0=0.05
-        )
+        parameterisation = pybop.XNES(cost=spm_costs, sigma0=0.05, max_iterations=125)
         parameterisation.set_max_unchanged_iterations(iterations=35, threshold=1e-5)
```
[the remark this answers is not in the document]
Comment (NicolaCourtier) [reply tests/integration/test_optimisation_options.py:85]: Sure, either is fine

File: pybop/optimisers/scipy_optimisers.py [hunk 34]
```diff
@@ -1,104 +1,193 @@
...
-        self.bounds = bounds
-        self.tol = tol
-        self.options = {}
-        self._max_iterations = maxiter
+    def __init__(self, cost, **optimiser_kwargs):
+        super().__init__(cost, **optimiser_kwargs)
 
-        if self.method is None:
-            self.method = "Nelder-Mead"
+    def _sanitise_inputs(self):
```
[the remark this answers is not in the document]
Comment (NicolaCourtier) [reply pybop/optimisers/scipy_optimisers.py:25]: I like the suggestion but let's keep separate for now. We can move in this direction in a future PR.

File: pybop/plotting/plot_convergence.py (hunk 32 above)
[the remark this answers is not in the document]
Comment (BradyPlanden) [reply pybop/plotting/plot_convergence.py:43]: Yes, I missed the `max()` in the lower condition. Here's the modified list comprehension version: 
```suggestion
  best_cost_per_iteration = [
      min((cost(solution) for solution in log_entry), default=np.inf)
      if optim._minimising
      else max((cost(solution) for solution in log_entry), default=-np.inf)
      for log_entry in log
]
```

File: examples/standalone/optimiser.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (BradyPlanden) [reply examples/standalone/optimiser.py:16]: Right, I think the standalone optimiser should be without the addition of a cost function within the class. Let's discuss this as a follow on though, as I realise it would require either skipping the file in the --examples tests or moving these standalones outside of the `examples' directory.

[the remark this answers is not in the document]
Comment (NicolaCourtier) [reply examples/standalone/optimiser.py:16]: I think it is consistent with the other Standalone examples for it to be defined this way, but yes, let's discuss further after this PR.

**Review by BradyPlanden — APPROVED** [review]
LGTM, thanks for the additions @NicolaCourtier!
